File: lintcode/wildcard-matching.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/wildcard-matching/

class Solution:
    def match_with_recursion(self, s, p):
        if len(s) == 0:
            if len(p) == 0:
                return True
            else:
                if p[0] == '*':
                    j = 1
                    while j < len(p) and p[j] == '*':
                        j += 1
                    if j < len(p):
                        return False
                    else:
                        return True
                else:
                    return False
        else:
            if len(p) == 0:
                return False
            else:
                if p[0] == '?' or p[0] == s[0]:
                    return self.isMatch(s[1:], p[1:])
                elif p[0] == '*':
                    j = 1
                    while j < len(p) and p[j] == '*':
                        j += 1
                    return self.isMatch(s, p[j:]) or self.isMatch(s[1:], p[j:]) or self.isMatch(s[1:], p)
                else:
                    return False

    def match_with_dp_with_2dmatrix(self, s, p):
        row = len(s)
        col = len(p)
        dp = [[False] * (col+1) for i in xrange(row+1)]

        dp[0][0] = True
        for i in xrange(1, row+1):
            dp[i][0] = False
        for j in xrange(1, col+1):
            dp[0][j] = p[j-1] == '*' and dp[0][j-1]

        for i in xrange(1, row+1):
            for j in xrange(1, col+1):
                if p[j-1] == '*':
                    dp[i][j] = dp[i-1][j-1] or dp[i-1][j] or dp[i][j-1]
                else:
                    dp[i][j] = dp[i-1][j-1] and (p[j-1] == '?' or p[j-1] == s[i-1])

        return dp[row][col]

    def match_with_backtrack(self, s, p):
        size_s = len(s)
        size_p = len(p)
        ptr_s, ptr_p = 0, 0
        saved_s, saved_p = None, None
        while ptr_s < size_s:
            if ptr_p < size_p and (s[ptr_s] == p[ptr_p] or p[ptr_p] == '?'):
                ptr_s, ptr_p = ptr_s+1, ptr_p+1
            elif ptr_p < size_p and p[ptr_p] == '*':
                saved_s, saved_p = ptr_s+1, ptr_p
                ptr_p += 1
            elif saved_s:
                ptr_s, ptr_p = saved_s, saved_p
            else:
                return False
        return p[ptr_p:].count('*') == size_p - ptr_p

    """
    @param s: A string
    @param p: A string includes "?" and "*"
    @return: A boolean
    """
    def isMatch(self, s, p):
        # return self.match_with_dp_with_2dmatrix(s, p)
        return self.match_with_backtrack(s, p)


if __name__ == '__main__':
    s = Solution()
    print s.isMatch("aab", "c*a*b")
    print s.isMatch("aa", "****a*****b***")
    print s.isMatch("aa", "*a*")
    print s.isMatch("aa", "*a**")
    print s.isMatch("aa", "*")
    print s.isMatch("abbabaaabbabbaababbabbbbbabbbabbbabaaaaababababbbabababaabbababaabbbbbbaaaabababbbaabbbbaabbbbababababbaabbaababaabbbababababbbbaaabbbbbabaaaabbababbbbaababaabbababbbbbababbbabaaaaaaaabbbbbaabaaababaaaabb", "**aa*****ba*a*bb**aa*ab****a*aaaaaa***a*aaaa**bbabb*b*b**aaaaaaaaa*a********ba*bbb***a*ba*bb*bb**a*b*bb")

File: leetcode/wildcard-matching.cc
// Question: https://leetcode.com/problems/wildcard-matching/

#include "headers.h"

class Solution {
public:
    bool isMatch(string s, string p) {
        // return backtrack(s, p);
        return dprog(s, p);
    }

private:
    bool dprog(string s, string p) {
        if (p.empty()) {
            return s.empty();
        }
        if (s.empty()) {
            return p.empty() || count(p.begin(), p.end(), '*') == p.size();
        }

        int row = s.size();
        int col = p.size();

        vector<vector<bool>> dp(row+1, vector<bool>(col+1, false));
        dp[0][0] = true;
        for (int i = 1; i < row+1; ++i) {
            dp[i][0] = false;
        }
        for (int j = 1; j < col+1; ++j) {
            dp[0][j] = dp[0][j-1] && (p[j-1] == '*');
        }

        for (int i = 1; i < row+1; ++i) {
            for (int j = 1; j < col+1; ++j) {
                if (p[j-1] == '*') {
                    dp[i][j] = dp[i-1][j] || dp[i][j-1] || dp[i-1][j-1];
                } else {
                    dp[i][j] = dp[i-1][j-1] && (s[i-1] == p[j-1] || p[j-1] == '?');
                }
            }
        }

        return dp[row][col];
    }

    bool backtrack(string s, string p) {
        int ps = 0, pp = 0, saved_ps = -1, saved_pp = -1;
        while (ps < s.size()) {
            if (pp < p.size() && (s[ps] == p[pp] || p[pp] == '?')) {
                ++ps;
                ++pp;
            } else if (pp < p.size() && p[pp] == '*') {
                saved_ps = ps+1;
                saved_pp = pp++;
            } else if (saved_ps != -1) {
                ps = saved_ps;
                pp = saved_pp;
            } else {
                return false;
            }
        }

        string subp = p.substr(pp);
        return (count(subp.begin(), subp.end(), '*') == subp.size());
    }
};

int main(int argc, char** argv) {
    Solution s;
    cout << s.isMatch("aa", "a") << endl;
    cout << s.isMatch("aa", "aa") << endl;
    cout << s.isMatch("aaa", "aa") << endl;
    cout << s.isMatch("aa", "*") << endl;
    cout << s.isMatch("aa", "a*") << endl;
    cout << s.isMatch("ab", "?*") << endl;
    cout << s.isMatch("aab", "c*a*b") << endl;
}
